cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java

```diff
@@ -183,8 +179,42 @@ private void checkService(Service headful) {
         assertEquals("TCP", headful.getSpec().getPorts().get(2).getProtocol());
         assertEquals(AbstractModel.METRICS_PORT_NAME, headful.getSpec().getPorts().get(3).getName());
         assertEquals(new Integer(KafkaCluster.METRICS_PORT), headful.getSpec().getPorts().get(3).getPort());
-        assertEquals("TCP", headful.getSpec().getPorts().get(2).getProtocol());
+        assertEquals("TCP", headful.getSpec().getPorts().get(3).getProtocol());
         assertEquals(kc.getPrometheusAnnotations(), headful.getMetadata().getAnnotations());
+
+        checkOwnerReference(kc.createOwnerReference(), headful);
+    }
+
+    @Test
+    public void testGenerateServiceWithoutMetrics() {
+        Kafka kafka = new KafkaBuilder(kafkaAssembly)
+                .editSpec()
+                    .editKafka()
+                        .withMetrics(null)
+                    .endKafka()
+                .endSpec()
+                .build();
+        KafkaCluster kc = KafkaCluster.fromCrd(kafka, VERSIONS);
+        Service headful = kc.generateService();
+
+        assertEquals("ClusterIP", headful.getSpec().getType());
+        assertEquals(expectedSelectorLabels(), headful.getSpec().getSelector());
+        assertEquals(3, headful.getSpec().getPorts().size());
+        assertEquals(KafkaCluster.REPLICATION_PORT_NAME, headful.getSpec().getPorts().get(0).getName());
+        assertEquals(new Integer(KafkaCluster.REPLICATION_PORT), headful.getSpec().getPorts().get(0).getPort());
+        assertEquals("TCP", headful.getSpec().getPorts().get(0).getProtocol());
+        assertEquals(KafkaCluster.CLIENT_PORT_NAME, headful.getSpec().getPorts().get(1).getName());
+        assertEquals(new Integer(KafkaCluster.CLIENT_PORT), headful.getSpec().getPorts().get(1).getPort());
+        assertEquals("TCP", headful.getSpec().getPorts().get(1).getProtocol());
+        assertEquals(KafkaCluster.CLIENT_TLS_PORT_NAME, headful.getSpec().getPorts().get(2).getName());
+        assertEquals(new Integer(KafkaCluster.CLIENT_TLS_PORT), headful.getSpec().getPorts().get(2).getPort());
+        assertEquals("TCP", headful.getSpec().getPorts().get(2).getProtocol());
```

Comment:
do we want to test that the metrics port is not there as well?

Reply:
It checks that there are 3 ports on line 202 and then checks the 3 ports one by one. So we know that the 4th port for metrics is not there.

Reply:
I am for an explicit test and not assuming it implicitly. But that is fine. Just merge.

Reply:
I'm not sure I understand how do you want to test it explicitly? 

The current tests doesn't make any assumptions - it makes sure there are only 3 ports and the 3 ports are plain, tls and replication. So there is 0 chance the metrics port is there.

Reply:
That's true. The ports are indexed and you cannot get anothe one. Merge! :)